packages/core/src/request.ts

```diff
@@ -131,6 +133,11 @@ export class Request<UserData extends Dictionary = Dictionary> {
      */
     handledAt?: string;
 
+    /**
+     * Local hooks for the request. Note that the hooks are not persisted once the request is stored to a storage.
+     */
+    hooks: Partial<Record<RequestEvent, ((request: Request) => void)[]>> = {};
```

Comment:
I agree with you that this is not ideal. The Request type is an amalgamation of multiple things and this adds one more. Also adding actual behavior to a previously more or less data-only class doesn't feel right.

I guess this is not a good fit for the `EventManager` that we already have, is it?

Reply:
Yeah, `Request` being data-only class worries me the most here, the hooks are a bit forced. Not sure about `EventManager`, right now it only handles Actor-wide events (`abort`, `migration` etc.), and is used as such... making it handle the crawling logic doesn't feel right either, mostly from the _what-depends-on-what_ POV.

In a perfect world, we would have the `ProxyConfiguration` predict the proxy tier from the request statelessly - the request contains the `sessionRotationCount`, (and can contain the `lastUsedProxyTier`), which is everything you need to determine the proxy - this is how it worked initially... until I realized that to go up one tier, you have to process the entire queue (of new requests) until you get to the retired request.

Reply:
Although - maybe `reclaimRequest` with `forefront` should make the crawler retry the request sooner? gotta need to try this, brb :)

Reply:
it works! 💥 check out the latest few commits - the way of predicting the proxy tier has changed completely (once again)